save-backend/src/main/kotlin/com/saveourtool/save/backend/utils/HttpUtils.kt

```diff
@@ -0,0 +1,147 @@
+@file:JvmName("HttpUtils")
+@file:Suppress("HEADER_MISSING_IN_NON_SINGLE_CLASS_FILE")
+
+package com.saveourtool.save.backend.utils
+
+import org.springframework.http.HttpHeaders
+import org.springframework.http.HttpHeaders.CACHE_CONTROL
+import org.springframework.http.HttpStatus.OK
+import org.springframework.http.ResponseEntity
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.InvocationKind.EXACTLY_ONCE
+import kotlin.contracts.contract
+import kotlin.system.measureNanoTime
+import kotlin.time.Duration.Companion.nanoseconds
+
+private const val SERVER_TIMING = "Server-Timing"
+
+/**
+ * [`no-transform`](https://www.rfc-editor.org/rfc/rfc7234#section-5.2.2.4)
+ * is absolutely necessary, so that SSE stream passes through the
+ * [proxy](https://github.com/chimurai/http-proxy-middleware) without
+ * [compression](https://github.com/expressjs/compression).
+ *
+ * Otherwise, the front-end receives all the events at once, and only
+ * after the response body is fully written.
+ *
+ * See
+ * [this comment](https://github.com/facebook/create-react-app/issues/7847#issuecomment-544715338)
+ * for details:
+ *
+ * The rest of the `Cache-Control` header is merely what _Spring_ sets by default.
+ */
+private val cacheControlValues: Array<out String> = arrayOf(
+    "no-cache",
+    "no-store",
+    "no-transform",
+    "max-age=0",
+    "must-revalidate",
+)
+
+/**
+ * Lazy HTTP response.
+ */
+typealias LazyResponse<T> = () -> T
+
+/**
+ * Lazy HTTP response with timings.
+ */
+typealias LazyResponseWithTiming<T> = () -> ResponseWithTiming<T>
+
+/**
+ * Adds support for the
+ * [`Server-Timing`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server-Timing)
+ * header.
+ */
+private fun <T : Any> T.withTimings(vararg timings: ServerTiming): ResponseWithTiming<T> =
+        ResponseWithTiming(this, *timings)
+
+/**
+ * Adds support for the
+ * [`Server-Timing`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server-Timing)
+ * header.
+ */
+private fun <T : Any> LazyResponse<T>.withTimings(vararg timings: ServerTiming): LazyResponseWithTiming<T> =
+        {
+            val response: T
+            val nanos = measureNanoTime {
+                response = this()
+            }
+
+            when {
+                timings.isEmpty() -> response.withTimings(
+                    ServerTiming(
+                        "total",
+                        "Total server time",
+                        nanos.nanoseconds
+                    )
+                )
+
+                else -> response.withTimings(*timings)
+            }
+        }
+
+/**
+ * @param timings the server-side timings.
+ * @param lazyResponse the lazy HTTP response.
+ * @return [lazyResponse] wrapped with HTTP headers.
+ */
+@OptIn(ExperimentalContracts::class)
+fun <T : Any> withHttpHeaders(
```

Comment:
Shouldn't `with*` function be an extension function?

Reply:
No, as it, as a side-effect, measures the execution time of the lambda it accepts and sets the `Server-Timing` header.

So it can be only made an extension on a lambda: `fun (() -> Response<T>).withHttpHeaders()`, which makes the function impractical.

Reply:
oh ok got it